src/group_theory/divisible.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import group_theory.subgroup.pointwise
+import group_theory.quotient_group
+import data.real.basic
+import data.complex.basic
+
+/-!
+# Divisible Group
+
+An abelian group `A` is said to be divisible iff for all `n ≠ 0` `y ∈ A`, there is an `x ∈ A`
+such that `n • x = y`. In this file, we adpot a constructive approach, i.e. we ask for an explicit
+`div_int : A → ℤ → A` function such that `div_int a 0 = 0` and `n • div_int a n = a` for `n ≠ 0`.
+
+## Main results
+* `add_comm_group.divisible_of_smul_surj` : the constructive definition of divisiblity is implied by
+  the condition that `n • x = a` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `add_comm_group.smul_surj_of_divisible` : the constructive definition of divisiblity implies
+  the condition that `n • x = a` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `add_comm_group.divisible_of_smul_top_eq_top` : the constructive definition of divisiblity is
+  implied by the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.smul_top_eq_top_of_divisible` : the constructive definition of divisiblity implies
+  the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.divisible_prod` : `A × B` is divisible for any divisible `A` and `B`.
+* `add_comm_group.divisible_pi` : Any product of divisble group is divisible.
+* `add_comm_group.divisible_of_char_zero` : Any field of characteristic zero is divisible.
+* `add_comm_group.divisible_quotient` : Quotient group of divisible group is divisible.
+* `add_comm_group.divisible_of_surj` : if `A` is divisible and `A →+ B` is surjective, then `B` is
+  divisible.
+TODO: Show that divisibility implies injectivity in the category of `AddCommGroup`.
+-/
+
+namespace add_monoid
+
+variables (A α : Type*) [add_monoid A] [has_scalar α A] [has_zero α]
+
+class divisible_by :=
+(div : A → α → A)
+(div_zero : ∀ a, div a 0 = 0)
+(div_cancel : ∀ {n : α} (a : A), n ≠ 0 → n • (div a n) = a)
+
+lemma smul_surj_of_divisible_by [divisible_by A α] {n : α} (hn : n ≠ 0) :
+  function.surjective ((•) n : A → A) :=
+λ x, ⟨divisible_by.div x n, divisible_by.div_cancel _ hn⟩
+
+
+end add_monoid
+
+namespace monoid
+
+variables (A α : Type*) [monoid A] [has_pow A α] [has_zero α]
+
+class rootable_by :=
+(root : A → α → A)
+(root_zero : ∀ a, root a 0 = 1)
+(root_pow : ∀ {n : α} (a : A), n ≠ 0 → (root a n)^n = a)
+
+attribute [to_additive rootable_by] add_monoid.divisible_by
+
+@[to_additive add_monoid.smul_surj_of_divisible_by]
+lemma root_surj_of_rootable_by [rootable_by A α] {n : α} (hn : n ≠ 0) :
+  function.surjective ((flip (^)) n : A → A) :=
+λ x, ⟨rootable_by.root x n, rootable_by.root_pow _ hn⟩
+
+end monoid
```

Comment:
@alexjbest I wasn't able to use `to_additive` to only write everything once. But is this what you envisioned, if so what should we call monoid that has root and in which file should we put the definition in?